### PR TITLE
Remote stream subscriber and replicator

### PR DESCRIPTION
**Pull request by grixa**
CC @dkorolev @mzhurovich

### CODE REVIEW

File: Sherlock/replicator.h [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef REMOTE_STREAM_REPLICATOR_H
```
Comment (mzhurovich) [inline Sherlock/replicator.h]: s/REMOTE_STREAM_REPLICATOR_H/CURRENT_SHERLOCK_REPLICATOR_H/

File: Sherlock/exceptions.h [hunk 2]
```diff
@@ -47,6 +47,30 @@ struct PublisherAlreadyOwnedException : SherlockException {
...
+
+struct SubscriptionIsNotActiveException : SherlockException {
+  using SherlockException::SherlockException;
+};
+
+struct RemoteStreamDoesNotRespondException : SherlockException {
+  using SherlockException::SherlockException;
+};
+
+struct RemoteStreamSchemaInvalidException : SherlockException {
```
Comment (dkorolev) [inline Sherlock/exceptions.h]: s/`SchemaInvalid`/`InvalidSchema`/g?

File: Sherlock/replicator.h [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+          internal_stop_(false),
+          remote_stream_(remote_stream,
+                         [this]() {
+                           TerminateSubscription();
+                           thread_.join();
+                         }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
```
Comment (dkorolev) [inline Sherlock/replicator.h]: This is a class invariant, not constructor invariant -- best to hoist it.

File: Sherlock/replicator.h [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+      bool terminate_sent = false;
+      while (!internal_stop_) {
+        if (!terminate_sent && external_stop_) {
+          terminate_sent = true;
+          if (subscriber_.Terminate() != ss::TerminationResponse::Wait) {
+            return;
+          }
+        }
+        try {
+          HTTP(ChunkedGET(
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Probably better to `CheckRemoteSchema()` here, prior to each `ChunkedGET` -- the source of data may restart.

File: Sherlock/replicator.h [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+      internal_stop_ =
+          (subscriber_(std::move(entry), idx, idxts_t() /*unused "last" arg*/) == ss::EntryResponse::Done);
+      std::unique_lock<std::mutex> lock(mutex_);
+      ++index_;
+      event_.notify_one();
+    }
+
+    void TerminateSubscription() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      event_.wait(lock, [this]() { return !subscription_id_.empty(); });
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `|| internal_stop_` in the condition?

File: Sherlock/replicator.h [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+      external_stop_ = true;
+      const std::string terminate_url =
```
Comment (dkorolev) [inline Sherlock/replicator.h]: And hide this into an `if (!subscription_id_.empty())`.

File: Sherlock/replicator.h [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+    void OnHeader(const std::string& header, const std::string& value) {
+      if (header == "X-Current-Stream-Subscription-Id") {
+        std::unique_lock<std::mutex> lock(mutex_);
+        subscription_id_ = value;
+        event_.notify_one();
+      }
+    }
+
+    void OnChunk(const std::string& chunk) {
```
Comment (mzhurovich) [inline Sherlock/replicator.h]: This implementation relies on the fact that each chunk is exactly one log entry, which in general is not true. I'd suggest to at least add a comment about this.

File: Sherlock/replicator.h (hunk 5 above)
Comment (mzhurovich) [inline Sherlock/replicator.h]: What happens to the `!subscription_id_.empty()` logic if the connection has been terminated abnormally?

File: Sherlock/replicator.h [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        return;
+      }
+
+      const auto split = current::strings::Split(chunk, '\t');
+      assert(split.size() == 2u);
+      const auto idx = ParseJSON<idxts_t>(split[0]);
+      assert(idx.index == index_);
+      auto entry = ParseJSON<TYPE_SUBSCRIBED_TO>(split[1]);
+      internal_stop_ =
+          (subscriber_(std::move(entry), idx, idxts_t() /*unused "last" arg*/) == ss::EntryResponse::Done);
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `/* */` comments are generally undesirable -- perhaps just `const idxts_t unused_idxts;` up there?

File: Sherlock/replicator.h [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+      std::unique_lock<std::mutex> lock(mutex_);
+      ++index_;
+      event_.notify_one();
```
Comment (mzhurovich) [inline Sherlock/replicator.h]: What's the purpose of this notification?

File: Sherlock/replicator.h [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+                                    const std::string& top_level_name,
+                                    const std::string& namespace_name)
+      : stream_(remote_stream_url),
+        schema_(
+            Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<T_STREAM_ENTRY>()).type_id,
+            top_level_name,
+            namespace_name) {
+    CheckRemoteSchema();
+  }
+  ~SubscribableRemoteStream() {}
```
Comment (dkorolev) [inline Sherlock/replicator.h]: This must have had some value :)

File: Sherlock/replicator.h (hunk 4 above)
Comment (grixa) [inline Sherlock/replicator.h]: Right. By the way, if the schema is changed, there will be a busy loop, is it OK?

Comment (dkorolev) [inline Sherlock/replicator.h]: A comment would do, for now. :-)

File: Sherlock/replicator.h [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    try {
+      return RemoteSubscriberScope<F, T_STREAM_ENTRY>(stream_, subscriber, start_idx, done_callback);
+    } catch (const current::sync::InDestructingModeException&) {
+      CURRENT_THROW(StreamInGracefulShutdownException());
+    }
+  }
+
+ private:
+  void CheckRemoteSchema() {
+    const auto response = HTTP(GET(stream_.ObjectAccessorDespitePossiblyDestructing().url_ + "/schema.simple"));
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Design suggestion for outside this pull request. Perhaps it's best to have a single "pre-subscribe" endpoint, would would:
1) Accept the desired schema as a header.
2) If the actual schema matches the desired one, pre-generate a subscription ID and return it, so that the actual subscription that follows next can request using the pre-generated ID instead of waiting for one in the header.
3) If the actual schema does not match, return both the 3-tuple schema signature and the actual schema, so that if we need to know what exactly is the difference between the schemas, the data to diff is right there.

File: Sherlock/replicator.h (hunk 5 above)
Comment (grixa) [inline Sherlock/replicator.h]: Then may be we should check `subscriber_thread_done_` here. But it also won't help if the thread would continuously fail to connect, for example.

File: Sherlock/replicator.h [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+    if (static_cast<int>(response.code) == 200) {
+      SubscribableSherlockSchema remote_schema =
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Don't these two fit one line?

File: Sherlock/replicator.h (hunk 5 above)
Comment (dkorolev) [inline Sherlock/replicator.h]: Yes, see the suggestion to pre-generate the subscription ID on the schema-validating request.

File: Sherlock/replicator.h [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+template <typename T_STREAM>
+struct StreamReplicatorImpl {
+  using EntryResponse = current::ss::EntryResponse;
+  using TerminationResponse = current::ss::TerminationResponse;
+  using entry_t = typename T_STREAM::entry_t;
+  using publisher_t = typename T_STREAM::publisher_t;
+
+  StreamReplicatorImpl(T_STREAM& stream) : stream_(stream) { stream.MovePublisherTo(*this); }
+  ~StreamReplicatorImpl() { stream_.AcquirePublisher(std::move(publisher_)); }
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `virtual` ?

File: Sherlock/replicator.h [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    assert(publisher_);
+    publisher_->Publish(entry, current.us);
+    return EntryResponse::More;
+  }
+
+  EntryResponse EntryResponseIfNoMorePassTypeFilter() const { return EntryResponse::More; }
+  TerminationResponse Terminate() const { return TerminationResponse::Terminate; }
+
+ private:
+  T_STREAM& stream_;
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Nit: `using stream_t` above as well, and `stream_t` here.

Also, upon introducing the `type_alias_t` semantics, we've moved from `T_STREAM` to just `STREAM` in template parameter names. Thanks!

File: Storage/test.cc [hunk 15]
```diff
@@ -1722,57 +1724,35 @@ TEST(TransactionalStorage, ReplicationViaHTTP) {
...
-  while (std::getline(body, line)) {
-    const size_t tab_pos = line.find('\t');
-    ASSERT_FALSE(tab_pos == std::string::npos);
-    const auto idx_ts = ParseJSON<idxts_t>(line.substr(0, tab_pos));
-    EXPECT_EQ(expected_index, idx_ts.index);
-    auto transaction = ParseJSON<Storage::transaction_t>(line.substr(tab_pos + 1));
-    ASSERT_NO_THROW(stream_publisher_owner.publisher_->Publish(std::move(transaction), idx_ts.us));
-    ++expected_index;
+  {
+    auto subscriber_scope = remote_stream.Subscribe(*replicator);
```
Comment (dkorolev) [inline Storage/test.cc]: `const`

File: Storage/test.cc [hunk 16]
```diff
@@ -1722,57 +1724,35 @@ TEST(TransactionalStorage, ReplicationViaHTTP) {
...
+    while (replicated_stream.InternalExposePersister().Size() < 2u) {
+      std::this_thread::yield();
```
Comment (dkorolev) [inline Storage/test.cc]: :+1:

**Comment by grixa** [comment]
It's not the final change yet

**Comment by grixa** [comment]
Almost done

**Comment by mzhurovich** [comment]
Looks good enough to merge, @dkorolev ?

**Comment by grixa** [comment]
Two more changes left: break ChunkedGET from itself using exception and use WaitableAtomic instead of condition_variable

**Comment by dkorolev** [comment]
Got it. Staying tuned!

**Comment by grixa** [comment]
Done.

File: Sherlock/exceptions.h [hunk 17]
```diff
@@ -47,6 +47,34 @@ struct PublisherAlreadyOwnedException : SherlockException {
   using SherlockException::SherlockException;
 };
 
+struct PublisherDoesNotExistException : SherlockException {
```
Comment (dkorolev) [inline Sherlock/exceptions.h]: Unused?

File: Sherlock/exceptions.h [hunk 18]
```diff
@@ -47,6 +47,34 @@ struct PublisherAlreadyOwnedException : SherlockException {
...
+  using SherlockException::SherlockException;
+};
+
+struct PublisherIsUsedException : SherlockException {
```
Comment (dkorolev) [inline Sherlock/exceptions.h]: Unused?

File: Sherlock/exceptions.h [hunk 19]
```diff
@@ -47,6 +47,34 @@ struct PublisherAlreadyOwnedException : SherlockException {
...
+  using SherlockException::SherlockException;
+};
+
+struct SubscriptionIsActiveException : SherlockException {
```
Comment (dkorolev) [inline Sherlock/exceptions.h]: Unused too. What am I missing?

File: Sherlock/exceptions.h [hunk 20]
```diff
@@ -47,6 +47,34 @@ struct PublisherAlreadyOwnedException : SherlockException {
...
+  using SherlockException::SherlockException;
+};
+
+struct SubscriptionIsNotActiveException : SherlockException {
```
Comment (dkorolev) [inline Sherlock/exceptions.h]: Unused :)

File: Sherlock/replicator.h [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+   public:
+    RemoteStream(const std::string& url, const std::string& top_level_name, const std::string& namespace_name)
+        : url_(url),
+          schema_(Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<entry_t>()).type_id,
+                  top_level_name,
+                  namespace_name) {}
+
+    void CheckSchema() const {
+      const auto response = HTTP(GET(url_ + "/schema.simple"));
+      if (static_cast<int>(response.code) == 200) {
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `static_cast<int>` is useful for unit tests to dump the "wrong" value as an integer. In real code, if (response.code == HTTPResponseCode.OK) is cleaner.

File: Sherlock/replicator.h [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+      } else {
+        CURRENT_THROW(RemoteStreamDoesNotRespondException());
+      }
+    }
+
+    const std::string GetURLToSubscribe(uint64_t index) const {
+      return url_ + "?i=" + current::ToString(index);
+    }
+
+    const std::string GetURLToTerminate(const std::string& subscription_id) {
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `) const {` ?

File: Sherlock/replicator.h [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+          remote_stream_(remote_stream, [this]() { TerminateSubscription(); }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
```
Comment (dkorolev) [inline Sherlock/replicator.h]: This needs a mutex, here and where it's accessed.

File: Sherlock/replicator.h [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+          TerminateSubscription();
+        }
+        thread_.join();
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Effectively, `!subscriber_thread_done_` is an invariant for "thread has not been joined yet", which is `thread_.joinable()`. Might be worth a comment or an assert.

File: Sherlock/replicator.h (hunk 22 above)
Comment (mzhurovich) [inline Sherlock/replicator.h]: What's the point of returning `const`?

File: Sherlock/replicator.h (hunk 24 above)
Comment (dkorolev) [inline Sherlock/replicator.h]: (Sorry for being picky here, just spent the morning making `Sherlock` and `Karl` tests stable. Thanks for understanding! :-) )

File: Sherlock/replicator.h [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+      } else {
+        if (done_callback_) {
+          done_callback_();
+        }
+      }
+    }
+
+   private:
+    static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Nit: a `static_assert` on a class-level invariant better be the first line of the body of the lass.

File: Sherlock/replicator.h [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+          }
+        }
+        try {
+          bare_stream.CheckSchema();
+          HTTP(ChunkedGET(
+              bare_stream.GetURLToSubscribe(index_),
+              [this](const std::string& header, const std::string& value) { OnHeader(header, value); },
+              [this](const std::string& chunk_body) { OnChunk(chunk_body); },
+              [this]() {}));
+        } catch (StreamTerminatedBySubscriber&) {
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `OnHeader()` here would save `subscription_id_`. Won't it make sense to erase it after `HTTP(ChunkedGET(...))` returns, as its value is no longer valid? (Or perhaps combine `subscription_id_` and `subscription_established_` in one `WaitableAtomic<>` would be a good idea?)

File: Sherlock/exceptions.h (hunk 19 above)
Comment (grixa) [inline Sherlock/exceptions.h]: Ah, right. These are reminders of an old version of the replicator.

File: Sherlock/replicator.h [hunk 27]
```diff
@@ -155,12 +155,12 @@ class SubscribableRemoteStream final {
...
         std::unique_lock<std::mutex> lock(mutex_);
         subscription_id_ = value;
-        event_.notify_one();
+        *subscription_established_.MutableScopedAccessor() = true;
       }
     }
 
     void OnChunk(const std::string& chunk) {
-      if (external_stop_ ) {
+      if (external_stop_) {
```
Comment (dkorolev) [inline Sherlock/replicator.h]: +1 to some `_requested` in the name.

File: Sherlock/replicator.h [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+
+    void OnChunk(const std::string& chunk) {
+      if (external_stop_) {
+        return;
+      }
+
+      // Expected one entire entry in each chunk,
+      // won't work in case of partial or multiple entries per chunk
+      const auto split = current::strings::Split(chunk, '\t');
+      assert(split.size() == 2u);
```
Comment (dkorolev) [inline Sherlock/replicator.h]: These should certainly not be `assert`-s -- why kill the whole binary when all it warrants is just HTTP connection re-initiation? After all, a TCP connection could be lost in the meantime, etc., etc. -- better be safe.

File: Sherlock/replicator.h [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+      const auto idxts = ParseJSON<idxts_t>(split[0]);
+      assert(idxts.index == index_);
+      auto entry = ParseJSON<TYPE_SUBSCRIBED_TO>(split[1]);
+      const idxts_t unused_idxts;
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Makes me wonder if it's default-initialized correctly -- I know it is, but looks unsafe. Why not make it a `const` member variable and initialize in constructor?

File: Sherlock/replicator.h [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+    void TerminateSubscription() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      subscription_established_.Wait(
+          [this](bool established) { return established || subscriber_thread_done_; });
+      external_stop_ = true;
+      if (!subscription_id_.empty()) {
+        const std::string terminate_url =
+            remote_stream_.ObjectAccessorDespitePossiblyDestructing().GetURLToTerminate(subscription_id_);
+        try {
+          const auto result = HTTP(GET(terminate_url));
```
Comment (dkorolev) [inline Sherlock/replicator.h]: lonely, lonely `result`, no one needs it :(

File: Sherlock/replicator.h [hunk 31]
```diff
@@ -0,0 +1,296 @@
...
+        } catch (current::net::NetworkException&) {
```
Comment (dkorolev) [inline Sherlock/replicator.h]: `current::Exception` instead of `current::net::NetworkException`?

File: Sherlock/replicator.h [hunk 32]
```diff
@@ -0,0 +1,296 @@
...
+        }
+      }
+    }
+
+   private:
+    bool valid_;
+    std::function<void()> done_callback_;
```
Comment (dkorolev) [inline Sherlock/replicator.h]: looks `const` to me

File: Sherlock/replicator.h [hunk 33]
```diff
@@ -0,0 +1,296 @@
...
+    using base_t = current::sherlock::SubscriberScope;
+
+   public:
+    using subscriber_thread_t = RemoteSubscriberThread<F, TYPE_SUBSCRIBED_TO>;
+
+    RemoteSubscriberScope(ScopeOwned<RemoteStream>& remote_stream,
+                          F& subscriber,
+                          uint64_t start_idx,
+                          std::function<void()> done_callback)
+        : base_t(std::move(
```
Comment (dkorolev) [inline Sherlock/replicator.h]: Just a thought: doesn't `base_t` have a perfect forwarding constructor (with `template<typename... ARGS>`)?

**Comment by dkorolev** [comment]
Looks great! Мы в офлайн часа на два с половиной, до вечера :)

File: Sherlock/replicator.h (hunk 23 above)
Comment (grixa) [inline Sherlock/replicator.h]: It is declared as `std::atomic_bool`. And what the point of acquiring the mutex here and below? It doesn't prevent us from skipping `subscriber_thread_done_` setting (and it's not bad). By the way, there is exact the same code in `sherlock::Stream::~SubscriberThreadInstance`. But I'm thinking in this case we can remove this check at all, because there is one more inside `TerminateSubscription()`. What do you think?

Comment (dkorolev) [inline Sherlock/replicator.h]: Agree -- and confirming Sherlock's code is battle-tested. The design you use is in fact very clean:

* `valid_` is initialized as `true` in constructor body, and would only be `false` if the constructor did throw -- `valid_` is pretty much semantically const.
* `subscriber_thread_done_`, on the other hand, only changes once, from `true` to `false`.

I'm concerned about the `subscription_established_.Wait([this](bool established) { return established || subscriber_thread_done_; });)` construct. Looks like both bad corner cases are possible:
1. If the server is not responding at all (which needs a test too), as internal termination has been triggered, the above`.Wait(...)`  may well wait forever, as it's not being notified.
2. On the other hand, the code in `TerminateSubscription()`, right after that `Wait()` might as well get executed twice: for instance, once from `ScopeOwnedBySomeoneElse()`'s callback, and one from this destructor. Also, as you don't clear `subscription_id_`, you may end up requesting to terminate twice (which is no harm at all, and your `TerminateSubscription()` ensures it's reentry-safe thanks to the top-level mutex lock, but still).

A better solution is a (much) longer stare and extensive tests for corner cases. A good enough solution, if you ask me, is to put more state into one `WaitableAtomic`, which would serve both as single state maintainer and the locking primitive to ensure 

Disclaimer: I'm no expert in multithreading; you may well have both more experience and have put more thought into   this snippet. Would be happy to be enlightened. Thank you! :innocent:

**Comment by dkorolev** [comment]
:+1:

**Comment by mzhurovich** [comment]
Ok, merging this and let's move further :)
